Bad Layer Handler Improvements

roya0045 opened the pull request:
## Description
This improves upon the previous work that I did for the bad feature handler.

This aims to add a button that will search for layers in an outward pattern up 4 levels (levels can be changed). The layers found will be highlighted in green. Then the user can hit apply or ok to confirm. If the folder containing the old path, it will move up until it reaches a valid folder and search from there. If the next valid folder is less than 4 levels away, it will also search up 4 levels, otherwise the search will be done only from the valid folder.

The layers in red will have to be changed through some other means.

The way alternative basepath are stored and handled has also changed. I have added existence check when attempting all alternative basepaths for a given original path.

This add a new button to launch the auto finder function. A query could be added to ask for the maximum depth to move up. In our current architecture 4 level is not much but for a user working in local, this might mean that the search could end up near the root and take a long time to find a file. 

Hence why I have added a way to limit vertical movement by ensuring that we are 3 folder away from the drive (by checking the number of system separator).

This is open for discussion.

Feature at work:
![autofinder](https://user-images.githubusercontent.com/12854129/66336231-b0671080-e90a-11e9-9a53-8b1b5edd1579.gif)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

roya0045 commented:
This is based on QEP 150.

roya0045 commented:
@nyalldawson Thanks for the tags

roya0045 commented:
@m-kuhn @nyalldawson Could one of you two add the squash tag, there are some commits remaining in this branch that I have no idea why they are there.  Mostly anything before 171971b.

roya0045 commented:
bumblefackery should be over on my end, @elpaso @nirvn  if you want to take a look at it.

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -364,79 +370,48 @@ void QgsHandleBadLayers::editAuthCfg()
+      fileName = longName.left( longName.lastIndexOf( '|' ) - 1 );
+    else
+      fileName = longName;
+    if ( !( item->foreground() == QBrush( Qt::green ) ) )
```
Is there a better way than checking the color? This is neither very readable nor very stable.

roya0045 replied:
I could simply add the layer by default but I want the user to validate if the layers that were found are the one that are desired. And not simply layer with the same name but the wrong content.

Otherwise I would have to add a registry if I want user validation.

roya0045 replied:
The goal is mostly to have user validation of the found file.

I could either merge it directly and bypass validation but this might lead to wrong input (if the file changed name or if there are duplicates).

Another way would be to maintain a registry of found file. But a more complex behaviour would have to be implemented to ask user for input to approve found layers.

Feedback from more devs would be appreciated on this aspect.

m-kuhn replied:
You can use `setData` to store the data separate from the representation:

https://doc.qt.io/qt-5/qtablewidgetitem.html#setData

Have a look at e.g. `QgsSettingsTree::Roles` to define custom roles.

roya0045 replied:
Couldn't I use Qt::ForegroundRole instead of defining a new role?

m-kuhn replied:
Better not, that will make it very hard for @nirvn to ship updates for his themes.

roya0045 replied:
Right, I'll see what I can come up with.

roya0045 replied:
What would be the optimal way to do things, the roles would be associated to the tableitem, now I'm not sure if Qt sets default values for every role. The reason being that if I declare a role in autofind and check for it in apply, there is a possibility that the user will only go for apply and if the desires role has not been set or has the wrong default value, this could have unintended consequences.

I'm not too sure if fiddling with roles and potentially uninitiated values is that much of a good idea.

I would assume that the color is carried as it can be assigned to a role. Since this role/foreground is only bound to the table item I'm missing the point as how this would affect the theme of qgis.  By using a default display role I have less of a risk to stumble upon a wrong value when checking the data.

nyalldawson replied:
set user data via setData is indeed the correct approach for something like this. The danger with using the colour is that it mixes presentation attributes with non-presentation, and it's almost certainly going to break in future when someone comes along and tweaks the shade of green used (not realising that you're relying on an exact colour check for other stuff elsewhere).

There's zero risk of uninitialised values. If the user data hasn't been set, you'll get an invalid QVariant value returned for it. These can be tested for by checking QVariant::isValid()

I'd suggest digging into qts docs on QVariant and the model/view approach. These are used everywhere throughout QGIS, so it's handy to know them well! 😀

roya0045 replied:
I have implemented something, using UserRole.

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +479,161 @@ int QgsHandleBadLayers::layerCount()
   return mLayerList->rowCount();
 }
+
+QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, const int driveMargin )
```
```suggestion
QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, int driveMargin )
```

... or is there a reason for this const? If not I'd remove it to be inline with the rest of our API.

roya0045 replied:
to prevent change during execution, mostly a safety measure. Should be removed now.

m-kuhn commented on src/app/qgshandlebadlayers.h:
```diff
@@ -63,18 +63,35 @@ class APP_EXPORT QgsHandleBadLayers
+     * Since: QGIS 3.10
+     */
+    QString checkBasepath( const QString &name, const QString &newPath, const QString &fileName );
+    /* Will check folder in an outward spiral up to 4 levels to check if the files exists
```
```suggestion
    /**
     * Will check folder in an outward spiral up to 4 levels to check if the files exists
```

roya0045 replied:
Thanks

m-kuhn commented on src/app/qgshandlebadlayers.h:
```diff
@@ -63,18 +63,35 @@ class APP_EXPORT QgsHandleBadLayers
 
     QString filename( int row );
     void setFilename( int row, const QString &filename );
+    /*checkBasepath will check if the newPath for the provided name is valid.
```
```suggestion
    /**
     * Checks if \a newPath for the provided \a name is valid.
```

It doesn't really mather a lot since it's private and in app. But if it's ever moved, it's nice if it's already properly formatted :)

roya0045 replied:
Thanks

m-kuhn commented on src/app/qgshandlebadlayers.h:
```diff
@@ -63,18 +63,35 @@ class APP_EXPORT QgsHandleBadLayers
+     * found files will be highlighted in green of approval, otherwise in red.
+     * maxDepth limit the number of time the search can move up
+     * driveMargin limits how close the search can get to the drive level
+     * Since: QGIS 3.10
```
```suggestion
     * \since QGIS 3.10
```

roya0045 replied:
Removed

m-kuhn commented on src/app/qgsguivectorlayertools.h:
```diff
@@ -81,3 +81,4 @@ class QgsGuiVectorLayerTools : public QgsVectorLayerTools
 };
 
 #endif // QGSGUIVECTORLAYERTOOLS_H
+
```
```suggestion
```

Try to avoid introducing changes in unrelated files.

roya0045 replied:
Side effect of reusing an old branch, should be fixed.

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +479,161 @@ int QgsHandleBadLayers::layerCount()
+    const QList<QString> altPaths = mAlternativeBasepaths.value( originalBase );
+    if ( ! altPaths.isEmpty() )
+    {
+      for ( QString altPath : altPaths )
```
```suggestion
      for ( const QString &altPath : altPaths )
```

roya0045 replied:
should be fixed

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +479,161 @@ int QgsHandleBadLayers::layerCount()
+  else if ( mAlternativeBasepaths.contains( originalBase ) )
+  {
+    const QList<QString> altPaths = mAlternativeBasepaths.value( originalBase );
+    if ( ! altPaths.isEmpty() )
```
No reason for this check I guess?

roya0045 replied:
Prevent iteration on an empty list.

m-kuhn replied:
Iterating an empty list will result in no iteration at all :)

roya0045 replied:
Very well, I just wanted to be more safe than segfault.

roya0045 commented:
@m-kuhn removing newline causes [code layout error](https://travis-ci.org/qgis/QGIS/jobs/549112552)
This is why the newline was added to the vectorlayertool.

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -117,8 +123,8 @@ QgsHandleBadLayers::QgsHandleBadLayers( const QList<QDomNode> &layers )
-    const QString basepath = datasource.left( datasource.lastIndexOf( '/' ) );
-    mFileBase[name].append( basepath );
+
+    mOriginalFileBase[name] = datasource.left( datasource.lastIndexOf( '/' ) );
```
I wonder if the layer id would be more suitable than the layer name for this registry?

roya0045 replied:
If they are maintained for each layer and robust, yes.

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -364,79 +370,48 @@ void QgsHandleBadLayers::editAuthCfg()
     QTableWidgetItem *item = mLayerList->item( i, 4 );
     QString datasource = item->text();
+    QString fileName;
+    const QString name { mLayerList->item( i, 0 )->text() };
```
Would it be possible to use the layer id instead of the name and save it as data (something like `mLayerList->item( i, 0 )->data(QgsHandleBasdLayers::LayerIdRole).toString()`)?

roya0045 replied:
Possibly, would this be robust enough to resist changes to the layer list? The idea is mostly to have the data associated to the layer and not have changes to the table (removing rows, etc) affect anything. If you tell me that the layerid will be maintained I don't have an issue with this.

m-kuhn replied:
In short yes.
Slightly longer: The `item` is created once in the code and then managed and moved around by the table (e.g. sorting) until the row is removed again. So any `item->setData` is safe and persistent. I'd always attach data that relates tot he whole row to the first column (`item( i , 0 )`) to make it easy to obtain again.

roya0045 replied:
Would `node.namedItem( QStringLiteral( "id" ) ).toElement().text()` be fine, as this is already in use.

m-kuhn replied:
Yes, that would be fine :+1:

roya0045 replied:
Keep in mind that `mapLayer->setDataSource( datasource, name, provider, options );` relies on the name and using the id forces me to use one more variable.

m-kuhn replied:
Ping @elpaso , I think you are most familiar with this code

nyalldawson replied:
You shouldn't need to change setDataSource here -- that doesn't require an ID, as the ID should never change (even after changing the layer's data source, it must still be left untouched)

roya0045 replied:
I don't plan to alter setDataSource, I simply wanted to say that relying on the name instead of the id is not removing the need to create the name variable.

roya0045 replied:
I have implemented something, using UserRole + 2 (as +0 is already used)

nyalldawson replied:
Ah, I misinterpreted, sorry

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +479,162 @@ int QgsHandleBadLayers::layerCount()
+  while ( !folder.exists() && folder.absolutePath().count( folder.separator() ) > driveMargin )
+  {
+    existingBase = folder.path();
+    if ( existingBase.contains( '/' ) )
```
Is this windows compatible?

roya0045 replied:
You are right, I'll use QDir::separator() instead, just to be safe.

m-kuhn replied:
Not sure if there's a risk that on Windows there will be `/` also. And if this can be avoided by using e.g. `absolutePath()`, would be worth a check (I don't have a windows system right now, so I can't sorry)

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +479,162 @@ int QgsHandleBadLayers::layerCount()
+    }
+  }
+
+  // Final cleanup: remove any bad layer (it should not be any btw)
```
Is this something that needs to be done in auto find?
I would imagine the workflow to be:

* click auto find
* hope it fixes things
* fix any leftovers manually

Is it me missing something or are all remaining layers (e.g. network based ones) removed as soon as auto find is clicked?

roya0045 replied:
I used the existing logic in apply() as a base. I assume that this could be removed as the cleanup will occur in apply or accept.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -85,10 +87,14 @@ QgsHandleBadLayers::QgsHandleBadLayers( const QList<QDomNode> &layers )
   mApplyButton = new QPushButton( tr( "Apply Changes" ) );
   mApplyButton->setToolTip( tr( "Apply fixes to unavailable layers (remaining unavailable layers will be removed from the project)." ) );
   buttonBox->addButton( mApplyButton, QDialogButtonBox::ActionRole );
+  mAutoFindButton = new QPushButton( tr( "Auto-Find layers" ) );
```
```suggestion
  mAutoFindButton = new QPushButton( tr( "Auto-Find Layers" ) );
```

roya0045 replied:
done, ty

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -85,10 +87,14 @@ QgsHandleBadLayers::QgsHandleBadLayers( const QList<QDomNode> &layers )
   mApplyButton->setToolTip( tr( "Apply fixes to unavailable layers (remaining unavailable layers will be removed from the project)." ) );
   buttonBox->addButton( mApplyButton, QDialogButtonBox::ActionRole );
+  mAutoFindButton = new QPushButton( tr( "Auto-Find layers" ) );
+  mAutoFindButton->setToolTip( tr( "Attempts to find the layers based on path name." ) );
```
```suggestion
  mAutoFindButton->setToolTip( tr( "Attempts to automatically find the layers based on path name." ) );
```

roya0045 replied:
done ty

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -113,12 +119,13 @@ QgsHandleBadLayers::QgsHandleBadLayers( const QList<QDomNode> &layers )
-    const QString basepath = datasource.left( datasource.lastIndexOf( '/' ) );
-    mFileBase[name].append( basepath );
+
+    mOriginalFileBase[id] = datasource.left( datasource.lastIndexOf( QDir::separator() ) );
```
To understand the logic here -- this is retrieving the folder from a folder + file path, right?

If so, better to use the inbuilt Qt methods for this -- see the QFileInfo class.

roya0045 replied:
You're right, should be fixed

roya0045 replied:
true, should be ok now.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -364,79 +371,48 @@ void QgsHandleBadLayers::editAuthCfg()
+    const QString layerId { node.namedItem( QStringLiteral( "id" ) ).toElement().text() };
+    const QString basepath = datasource.left( datasource.lastIndexOf( QDir::separator() ) );
+    const QString longName = datasource.mid( datasource.lastIndexOf( QDir::separator() ) );
+    if ( longName.lastIndexOf( '|' ) != -1 )
```
Instead of this approach, it would be much safer to use QgsProviderRegistry::decodeUri and extract the path component from the results -- that automatically handles stripping any layerName parameters. (Not all providers use "|" to identify these)

roya0045 replied:
I have a question, would this work even on invalid files. My aim is simply to remove the extra arguments passed, in order to only search for the file by name and then append them back to the filepath.

 Now when I look at the  provideregistry::decideUri (at least in python) I need to make an instance of QgsProviderRegistry as the function is not static. This means that I must have a valid filepath to make the providerregistry, I'm also not even sure what is going to be outputed at this point.

nyalldawson replied:
Yes -- it'll work on all layers, because it delegates to the provider itself to interpret the path exactly as it would if it was trying to load the layer.

You would use QgsProviderRegistry::instance()->decodeUri( layer_uri, layer_provider_name ), and then test if the returned map includes a "path" entry. If it doesn't, it's not file based (maybe db, maybe url, ...). If it does, then that's the path this dialog needs to work with.

roya0045 replied:
Great, thanks a ton!

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +480,143 @@ int QgsHandleBadLayers::layerCount()
+  int depth = 0;
+  QDir folder = QDir( basePath );
+  QString existingBase;
+  while ( !folder.exists() && folder.absolutePath().count( folder.separator() ) > driveMargin )
```
folder.absolutePath().count( folder.separator() ) is fragile -- use QDir/QFileInfo classes to iterate up the directory chain instead.

roya0045 replied:
How would you suggest to validate the depth of the current folder as I couldn't find an alternative way to check the depth of the data. This is important as I don't want to get close to the root to perform search, this would stall the system and possibly provide wrong results.

nyalldawson replied:
I think in this case it might be worth iterating up from the original path until there's no parents directories left (or a suitable cutoff is hit... e.g. 20 folders), and building this list in its entirety. Then you could trim off the last n entries from the list, where n == the minimum depth you want to allow from a root folder.

roya0045 replied:
Wouldn't that mean looping twice on things? Also QDir cannot cdUp if the directory does not exist, this would mean that I have to fiddle as I do when attempting to find the first existing directory. I'm not sure how optimal this is compared to checking for system separators. If you tell me that this is the only way, then I'm ok with it. It just might be a hassle to build depth index, keep the related path in order to properly set a stop when the path is encountered. :/

roya0045 replied:
I'll await your confirmation if you truly want me to loop over thing instead of using the / separator that is used in QT. I feel like the looper that you proposed would provide an useless overhead (in term of code, and runtime) when relying on the unix separator is robust.

nyalldawson replied:
The tricky thing here is.... we can't gaurantee that projects use the correct slash for their current operating system :scream: There's potential for projects to be shared between Windows and MacOS, for instance. Even on a single operating system environment (linux), I still see windows style paths creep into layer sources in some cases.

It makes things super tricky! I wasn't aware that QDir couldn't iterate up through non-existing paths - that's unfortunate.

I'm not sure what the best approach forward is here. Maybe we should treat either \ or / the same on any platform, and consider them both as candidates for a new folder. What do you think? @m-kuhn any additional thoughts?

m-kuhn replied:
> The tricky thing here is.... we can't gaurantee that projects use the correct slash for their current operating system scream There's potential for projects to be shared between Windows and MacOS, for instance. Even on a single operating system environment (linux), I still see windows style paths creep into layer sources in some cases.

Is this true? Wouldn't this lead to issue trying to open a Windows project under *nix systems?

> I'm not sure what the best approach forward is here. Maybe we should treat either \ or / the same on any platform, and consider them both as candidates for a new folder.

... but if it is, the approach of treating `\` and `/` equally everywhere sounds like the best bet to me.

roya0045 replied:
From what I gather (I might be wrong) if we rely on QDir/QFileInfo to save and parse filepath, this would ensure that every path is in the same format (in this instance the unix format), otherwise I would also assume that QDir and QFileInfo would do the proper conversion when receiving a path that is formatted for another system. ( Could anyone do a quick check on linux with a windows formatted path to validate this please?)

Since all paths would be converted to the same formatting, I would then assume that it is appropriate to relying on the unix separator in order to obtain folder depth.

roya0045 commented:
@nyalldawson you are right, using QDir::seperator() is bad since it returns the OS separator but all path are converted to unix style with '/' separator. This is why I have remove all QDir::sparator() and am now using '/' to count the depth. I cannot test on mac but I assume that paths in qt on mac also have / as folder separator. This ensure that depth checks are always valid.

roya0045 commented:
@nyalldawson are you ok with the use of the unix separator to estimate depth given the fact that Qt convert all path to unix style?

roya0045 commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +496,154 @@ int QgsHandleBadLayers::layerCount()
+  QDir folder = QDir( QDir( basePath ).absolutePath() );
+  QString existingBase;
+  // find the nearest existing folder
+  while ( !folder.exists() && folder.absolutePath().count( '/' ) > driveMargin )
```
@nyalldawson Are you now ok with this approach or was there something that you wanted me to change?

nyalldawson replied:
All good!

roya0045 commented:
Ready for merge until further notice.

roya0045 commented:
@nyalldawson @m-kuhn  since we don't use qt to deal with paths, I have added a function to convert back the path to the local system style. This means that counting depth using unix seperators is still ok but the output will be properly formatted and can handled by the proper provider.

This was an oversight on my end but things should be all good and ready to merge now.

roya0045 commented:
@nyalldawson @m-kuhn I have done some cleanups in the commits, some remain but I think this should be ok to merge. Let me know if there is anything else.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+  while ( !folder.exists() && folder.absolutePath().count( '/' ) > driveMargin )
+  {
+    existingBase = folder.path();
+    if ( !folder.cdUp() )
```
I'm unsure, but it looks like this could potentially get stuck in a endless loop if cdUp fails. Also - what happens with symbolic links here? I think it needs an iteration counter where it bombs out after maxDepth iterations.

roya0045 replied:
Are you suggesting to also use the depth index as a limiter? Since one count in term of movement and the other in terms of absolute depth this could be done but I didn't want to limit the search possibilities when there are major overhauls in a folder structure. This is a balance that I find hard to achieve with two limiting elements. I'll try to restructure things.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+    if ( !folder.cdUp() )
+      folder = QFileInfo( existingBase ).absoluteDir(); // using fileinfo to move up one level
+    depth += 1;
+  }
```
I'm a bit shaky on the logic here -- can you confirm I have it correct?
- at this stage we've trimmed subfolders off the original folder path until we've eventually hit a folder which actually still exists (and I think we should be returning out early here if that didn't happen)
- we then keep traversing up through the folders, at each stage searching through all subdirectories looking for the file?

If so, I can understand the intention. But my concern is that we're duplicating searches, since every time we move up a folder we're also re-checking the subfolder (and it's contents) which we just came from.

roya0045 replied:
Yes this is the case and I agree that there is some wasted potential, it might be possible to skip the folder that was already 'scanned' but I would need to add another loop for that if you don't mind.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
   return mLayerList->rowCount();
 }
+
+QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, int driveMargin )
```
This function is quite general (and potentially useful outside of bad layer handler). Could you move it to QgsFileUtils?

roya0045 replied:
Absolutely, once this is fine I'll migrate.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+
+    if ( !( dataSourceChanged ) )
+    {
+      datasource = QDir::toNativeSeparators( findFile( fileName, basepath ).replace( fileName, longName ) );
```
again - better remove the toNativeSeparators here. (sorry if that was my fault!).

roya0045 replied:
Please see previous comment, as to why this is important.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+
+    if ( !( dataSourceChanged ) )
+    {
+      datasource = QDir::toNativeSeparators( findFile( fileName, basepath ).replace( fileName, longName ) );
```
What's the .replace( fileName, longName ) do?

roya0045 replied:
Ah, this is because when displayed the layers such as local DB (gpkg, sqlite,etc) have extra elements to fetch the proper layer and those extra elements are not good when trying to find the file. The goal is to replace the name of the file with the name+query combo in order to be processed properly once the file is found.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+{
+  const QString originalBase = mOriginalFileBase.value( layerId );
+  const QFileInfo newpathInfo = QFileInfo( newPath );
+  if ( newpathInfo.exists() && newpathInfo.isFile() )
```
Can you explain the logic regarding mAlternativeBasepaths for me?

roya0045 replied:
Ah yes, the logic is based on my first PR about the handler. I am assuming that if someone is moving folders entirely then when many of those layer break the changes that occured for one layer should be the same for every other layer.

This way I aimed to make a dictionary of viable paths for each basepath. IE if something in a given folder is split in 3 distinct folders, than the original path will be associated with the three new viable paths that are encountered and those are reused to speed up fetching.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -364,79 +371,63 @@ void QgsHandleBadLayers::editAuthCfg()
+    }
 
-    if ( mFileBase[ name ].size() == 1 )
+    QVariantMap providerMap = QgsProviderRegistry::instance()->decodeUri( dataInfo.absoluteFilePath(), provider );
```
```suggestion
    QVariantMap providerMap = QgsProviderRegistry::instance()->decodeUri( provider, dataInfo.absoluteFilePath() );
```

roya0045 replied:
Thanks, should be fixed!

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+{
+  QDir::setCurrent( QgsProject::instance()->absolutePath() );
+  QgsProject::instance()->layerTreeRegistryBridge()->setEnabled( true );
+  buttonBox->button( QDialogButtonBox::Ignore )->setEnabled( false );
```
Remove

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+        provider = "ogr";
+    }
+
+    QVariantMap providerMap = QgsProviderRegistry::instance()->decodeUri( dataInfo.absoluteFilePath(), provider );
```
```suggestion
    QVariantMap providerMap = QgsProviderRegistry::instance()->decodeUri( provider, dataInfo.absoluteFilePath() );
```

roya0045 replied:
done thanks.

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+  }
+  if ( depth > maxDepth )
+    maxDepth = depth;
+  while ( depth <= maxDepth && folder.cdUp() && folder.absolutePath().count( '/' ) > driveMargin )
```
I think the cdUp should be deferred until this loop has run at least once -- otherwise if the move up here results in the driveMargin check failing on the first iteration, you skip checking the actual first found matching folder.

roya0045 replied:
Good point, the idea of the limit was to prevent the searching function to go through most of the server/system and bog things down, the limit of 3 level was something that I found convenient for us.

roya0045 replied:
I have changed the logic, let me know what you think.

nyalldawson commented:
@roya0045 I've been testing this branch in practice, and with the above commented fixes I can get the auto-find functionality to work and determine the correct path for moved files (they turn green). But then clicking "apply" has no effect, and I can't get the changes to actually save.

I'm wondering why the changes in the apply routine are needed here at all. Couldn't the auto-find functionality work without modification to the apply? I.e., it searches, fixes the paths, and then leaves the corrected paths in the table cell, and then they'll apply like normal using the existing apply method...

roya0045 commented:
> @roya0045 I've been testing this branch in practice, and with the above commented fixes I can get the auto-find functionality to work and determine the correct path for moved files (they turn green). But then clicking "apply" has no effect, and I can't get the changes to actually save.
> 
> I'm wondering why the changes in the apply routine are needed here at all. Couldn't the auto-find functionality work without modification to the apply? I.e., it searches, fixes the paths, and then leaves the corrected paths in the table cell, and then they'll apply like normal using the existing apply method...

This is what I was going for but I also wanted to highlight the files found so that the user can validate them, it is possible that I have done something that does not work out well. I'll try to find what might go wrong.

Thanks for the feedback!

roya0045 commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+      continue;
+    }
+
+    datasource = QDir::toNativeSeparators( checkBasepath( layerId, basepath, fileName ).replace( fileName, longName ) );
```
The reason for the basepath are because while QDir will convert to unix, our provider will only accept paths with native separators, this is why I want to recast those string in a 'native' style to prevent breakage when it comes the time to import the actual layer.

roya0045 replied:
I highly disagree for the following reason: our providers need the properly formatted paths. The conversion to native separators is to ensure that the proper path is stored and used by our provider since only QDir and such are OS independent, but not our providers. Not doing this conversion would lead to unix style path on every system and would not be usable on windows.

roya0045 commented:
@nyalldawson Concerning the changes in the apply function that relate to the user role, the goal was to check and log the alternative path and without having to do it on every layers, it was meant to use the layers that were found in order to complete the 'registry' of path changes. This is needed as the find function can be done on a single or group of layers and then the results can be leveraged to potentially find other layers when the user hits apply.

roya0045 commented:
@nyalldawson forgot to save before commit and uploaded partial changes, should be fine for review now.

roya0045 commented:
@nyalldawson do you have a clue about the failing test, it seems to be a desired failure but it fails the wrong way and I doubt my changes are related. This is strange and I'm thinking about disabling the test.

nyalldawson commented:
Does it work locally?

roya0045 commented:
Didn't get the time to build locally, but I'm confused since no changes that I made should affect such a random test, especially since it passed previously.

nyalldawson commented:
You've **GOTTA** get your local build sorted - it's not appropriate to rely on existing unit tests alone to test new functionality, there's a responsibility that contributors have done due diligence and tested well locally before submission. 

Grab VirtualBox, a Ubuntu iso, and in no time you'll have a working local build environment. And you'll save hours and hours if workarounds as a result 😁

roya0045 commented:
> You've **GOTTA** get your local build sorted - it's not appropriate to rely on existing unit tests alone to test new functionality, there's a responsibility that contributors have done due diligence and tested well locally before submission.
> 
> Grab VirtualBox, a Ubuntu iso, and in no time you'll have a working local build environment. And you'll save hours and hours of workarounds as a result 😁

Oh don't worry this is what I have done, its simply that on my home setup the first build takes 8 hours on my vm and I need to work on other projects that need the resources, hence why I cannot build things soon. I might get a vm at work to help speed things up but I still have some hurdles to cross beforehand.

roya0045 commented:
I might be able to test by the end of the week or early next week.

roya0045 commented:
@nyalldawson I should be able to build and test in the coming week.

roya0045 commented:
@nyalldawson Ok I see what you mean, somehow the path becomes green but it isn't changed, I have also noticed an issue with some paths where the wrong file can be found. I'll work on those issues.

roya0045 commented:
@nyalldawson I'm trying to improve some aspects and I'm stuck wondering what to do when many results are found. Currently I only return the first result found.

This mean that if every folder contains 1.tif and if the desired folder is renamed, the first folder containing 1.tif will be returned. I have been looking into a way to address this issue.

First I thought of implementing some means of fuzzy matching using string distance but qt doesn't have something that can do this. And I don't want to add a string distance algorithm.

I'm working on a popup window that lists found layers and asks the user to select the proper source. Though I'm not sure how to link everything from the ui file. I'll upload the broken build as a backup.

roya0045 commented:
I'm not sure how to deal with the ui file given that other files seems to be imported fine as a ui_*.h element but this one doesn't. Do I need to perform something to fix this issue?

roya0045 commented:
@nirvn would you happen to know what is the issue when I'm attempting to import a new UI file?

 I assumed that the ui_*.h files were generated automatically since there were none in qgis but somehow the include fails stating that the header does not exist. Do I need to run a script or change something somewhere to make sure that the header of the ui is generated?

nirvn commented on src/core/qgsfileutils.h:
```diff
@@ -87,6 +90,58 @@ class CORE_EXPORT QgsFileUtils
+    /**
+     * Constructor for the list selector.
+     * \param items List of items to display
+     * \label Label to display above the list of items
```
Your missing the param keyword

roya0045 replied:
thanks!

nirvn commented:
Alex , change the class name in your .ui file, run cmake, then make. It should work.

roya0045 commented:
> Alex , change the class name in your .ui file, run cmake, then make. It should work.

Dammit, Thanks a ton for the help!

roya0045 commented on src/ui/qgslistselectorbase.ui:
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
+ <class>dialog</class>
```
I never thought this parameter could be so important...

roya0045 commented:
@nirvn I still can't get it to work locally or in travis. But I'm starting to wonder if we already have a generic window where a user can select something. Something like qgsdatasourceselectdialog but without the filter, the buttons and just a label (to customize). 
I'm assuming I could alter the widget in my code, but this might be harder to debug.

roya0045 commented:
@nyalldawson if I want the user to select an item from the list of possible paths when more than one result is found, what is the best way to go on about this?

I have created a new ui but the include statement doesn't work and I feel that this thing should either be a generic widget to be reused or adapt an existing widget (but none seem to be close to what I want). Should I make the widget standalone or is there a better option to get the user to select an item from a list?

nyalldawson commented:
@roya0045 You may be able to reuse QInputDialog::getItem ? https://doc.qt.io/qt-5/qinputdialog.html#getItem

roya0045 commented:
> @roya0045 You may be able to reuse QInputDialog::getItem ? https://doc.qt.io/qt-5/qinputdialog.html#getItem

How come those type of results never pop up when I'm using a search engine or searching in Qt.. 

This should do the trick and save me some more headache!

roya0045 commented:
@nyalldawson @m-kuhn did a live test, worked perfectly on windows. Search time can be slow but this might be due to our architecture/server size. 

This is ready to merge IMO!

roya0045 commented:
@nyalldawson @m-kuhn Did some more test, some thing weird is going on and it sometimes try to open the bad path and fails. I'll try to solve it when I got some time, otherwise the UI part is ok.

roya0045 commented:
@nyalldawson @m-kuhn Ok found the issue, allowed search to be one level deeper.

I have added a gif to display the plugin in action.

Its ok to merge IMO, BUT there is one thing that can be improved, when using the layer finder, there is a CRS selection popup, but when applying the changes and when the layers are added to the projet, the CRS selection popup appears again.

Is there a way to have the CRS selector pop only once?

roya0045 commented:
@nyalldawson @m-kuhn, any chances to get this ' merge after thaw' ? The feature works as intended and kinks should have been ironed out. Unless I need to add a skip/check for web layers.

roya0045 commented:
Web layers are already skipped, forgot about that. All should be ok then.

roya0045 commented:
@m-kuhn @nyalldawson Simple ping for a review/approval when possible. Everything should work as intended.

m-kuhn commented:
@elpaso is there a (partial) overlap with your resent resolve path functionality?

m-kuhn commented:
But the base concept is the same: finding tables that have gone missing?
Isn't there the potential for a shared approach that helps both sides?

elpaso commented:
Not really: missing dependencies are not like broken layers, are more like good layers that have not been loaded, the only partially overlapping case is the changed schema for DBs.

But I was also thinking to a generic approach, my idea (not yet funded!) is to have a registry of `QgsProviderLayerFinder` that a provider registers to an application level registry.
The provider knows how to search for layers and can be asked to find a missing layer, whether is is a broken layer or it has just not being loaded.

But I'm not sure it fits well in case of broken DB layers: my PR is searching for layers in the "good" existing connection of the same "good" layer that requires a dependency: the main use case for value-relations is typically within the same DB (or you would not have any referential integrity).

In case of broken layers, you typically have a broken connection (except for the changed schema case), in that case the QgsProviderLayerFinder would not help because it has no knowledge about not-registered connections.

Just to make things more complex, it would be nice to have a similar system for broken relations and joins.

m-kuhn commented:
I have seen people use ValueRelations with shapefiles and xls too :man_shrugging: 

@roya0045 can you remind us in a few words which usecase you actually target here. I.e. not how it works (that's described in the topmost post), but how does a user get into the situation that he requires this particular functionality.

roya0045 commented:
> I have seen people use ValueRelations with shapefiles and xls too 🤷‍♂
> 
> @roya0045 can you remind us in a few words which usecase you actually target here. I.e. not how it works (that's described in the topmost post), but how does a user get into the situation that he requires this particular functionality.

The main use case is to fix project with multiple broken dependencies. If we move the project file and have relative paths, or restructure our folders, change the names of the folders, move layers, etc. This aims to quickly patch up projects that now only have broken layers due to change in our architecture instead of having to find the file manually with browse or by copy pasting the proper path this just finds the closest matches and patch up project. This way old projects can be quickly patched and used again.

Is this satisfactory?

roya0045 commented:
@m-kuhn My initial PR for the bad layer handler was in the same spirits as Esri's fixing behavior but this one aims to make things more trivial at the expense of some runtime while leveraging my previous PR.

This is mostly for users dealing with local files/db and not servers.

m-kuhn commented:
So, the problem statement would be:

> User moved his project file (with relative paths) around or moved the data around.

And the approach is:

> Offer the use to "auto find" his data, assuming that the data is somewhere "close" to the project file.

I can't remind when that happened to me last time, but I assume it happened somewhere close to you and that's why you propose this :)

What's the definition of close again? Up to 4 hierarchy levels higher than the project file (plus all subfolders in these 4 levels"? And into subfolders in the project file's folder as well?

m-kuhn commented on src/core/qgsfileutils.h:
```diff
@@ -87,6 +87,20 @@ class CORE_EXPORT QgsFileUtils
     static QString findClosestExistingPath( const QString &path );
+
+    /**
+     * Will check folder in an outward spiral up to 4 levels to check if the files exists
```
```suggestion
     * Will check \a basepath in an outward spiral up to 4 levels to check if \a fileName exists
```

roya0045 replied:
Thanks!

m-kuhn commented on src/core/qgsfileutils.h:
```diff
@@ -87,6 +87,20 @@ class CORE_EXPORT QgsFileUtils
+
+    /**
+     * Will check folder in an outward spiral up to 4 levels to check if the files exists
+     * found files will be highlighted in green of approval, otherwise in red.
```
This method does not colorize anything, does it?

roya0045 replied:
Yes it does, does that prove to be a problem?

roya0045 replied:
Ah no sorry, this was supposed to be in the doc for the bad layer handler, not this. My bad.

m-kuhn commented:
Ok, convinced. Looks like a good thing to do.
Can you add a couple of tests for `QgsFileUtils::findFile` to make sure it works (and demonstrate what it finds)?

- Original file exists
- File exists in higher levels (less nesting)
- File exists in lower levels (more nesting)
- Provided file path is invalid
- No result found
- Max depth exceeded by 1 (no result found)
- Multiple matching files in different hierarchies exist (order of precedence)

Hint: data driven testing should make it fairly easy to write these tests efficiently. https://doc.qt.io/qt-5/qtest.html#QFETCH but no obligation to use it

roya0045 commented:
> So, the problem statement would be:
> 
> > User moved his project file (with relative paths) around or moved the data around.
> 
> And the approach is:
> 
> > Offer the use to "auto find" his data, assuming that the data is somewhere "close" to the the original data path.
> 
> [Edit: corrected with information from first message]

Exactly!

> Ok, convinced. Looks like a good thing to do.
> Can you add a couple of tests for `QgsFileUtils::findFile` to make sure it works (and demonstrate what it finds)?
> 
> * Original file exists
> * File exists in higher levels (less nesting)
> * File exists in lower levels (more nesting)
> * Provided file path is invalid
> * No result found
> * Max depth exceeded by 1 (no result found)
> * Multiple matching files in different hierarchies exist (order of precedence)
> 
> Hint: data driven testing should make it fairly easy to write these tests efficiently. https://doc.qt.io/qt-5/qtest.html#QFETCH but no obligation to use it

I'll see what I can do, though for your last point, since it goes outward it'll return the first match encountered. Though since QInputDialog is used, I'm not too sure how to deal with that for such test, I think I would rather skip that case.

m-kuhn commented:
> Though since QInputDialog is used, I'm not too sure how to deal with that for such test, I think I would rather skip that case.

Can you move the input dialog to a different level?
Ideally core should not use any widgets so it can be used on server/mobile qml based apps/ background threads too.
I'd probably return a list and let the caller decide what to do, but some smart callback based approach might work too. In case this is not possible, at least mention it in the api docs.

roya0045 commented:
> > Though since QInputDialog is used, I'm not too sure how to deal with that for such test, I think I would rather skip that case.
> 
> Can you move the input dialog to a different level?
> Ideally core should not use any widgets so it can be used on server/mobile qml based apps/ background threads too.
> I'd probably return a list and let the caller decide what to do, but some smart callback based approach might work too. In case this is not possible, at least mention it in the api docs.

I think I'll go for a list and handle it in the bad layer handler part.

roya0045 commented:
@m-kuhn took me some time to get everything sorted out, I made more tweaks/improvements to the finders and tests are all good now.

Should be mergeable.

roya0045 commented:
@m-kuhn The name for the dir in the tests might not be optimal but I think that a review of the tests should help clear up the current behavior. Yes it returns a file path even if the path points to a non-existing file. I am debating this aspect of return a path to a non-existing file or an empty string in the list of string /  empty list.

roya0045 commented:
Decided to return a list with an empty string when nothing found, made more sense and would be easier to deal with.

m-kuhn commented:
Sorry for the late reply. I think an empty list would be better suited. (i.e. "no item found", whereas a list with one empty item means more something like "one item found with an empty path").

roya0045 commented:
> Sorry for the late reply. I think an empty list would be better suited. (i.e. "no item found", whereas a list with one empty item means more something like "one item found with an empty path").

Understood, code changed. Thanks for the feedback!

m-kuhn commented on src/core/qgsfileutils.h:
```diff
@@ -87,6 +87,19 @@ class CORE_EXPORT QgsFileUtils
+     * \returns List of strings of the first matching path in unix format.
+     * \since QGIS 3.12
+     */
+    static QList<QString> findFile( const QString &fileName, const QString &basepath, int maxDepth = 4, int driveMargin = 3, const QString &currentDir = "" );
```
```suggestion
    static QStringList findFile( const QString &fileName, const QString &basepath, int maxDepth = 4, int driveMargin = 3, const QString &currentDir = "" );
```

this is a drop in replacement for `QList<QString>` (inherits from it) and has some nice additions.

roya0045 replied:
I always though it was an alias.

m-kuhn commented on src/core/qgsfileutils.cpp:
```diff
@@ -19,6 +19,9 @@
 #include <QSet>
+#include <QDirIterator>
+#include <QWidget>
+#include <QDebug>
```
```suggestion
```

roya0045 replied:
ty

m-kuhn commented on src/core/qgsfileutils.cpp:
```diff
@@ -19,6 +19,9 @@
 #include <QDir>
 #include <QSet>
+#include <QDirIterator>
+#include <QWidget>
```
```suggestion
```

roya0045 replied:
ty

m-kuhn commented on src/app/qgshandlebadlayers.h:
```diff
@@ -64,17 +64,36 @@ class APP_EXPORT QgsHandleBadLayers
+    // Registry of the original paths associated with a file as a backup
+    QHash <QString, QString > mOriginalFileBase;
+    // Keeps a registry of valid alternatives for a basepath
+    QHash <QString, QList<QString> > mAlternativeBasepaths;
```
```suggestion
    QHash <QString, QStringList > mAlternativeBasepaths;
```

roya0045 replied:
done ty.

m-kuhn commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -450,16 +444,11 @@ void QgsHandleBadLayers::apply()
 
-  // Final cleanup: remove any remaining bad layer
-  // (there should not be any at this point)
+  // Final cleanup: remove any bad layer (it should not be any btw)
```
```suggestion
  // Final cleanup: remove any bad layer (it should not be any by now)
```

roya0045 replied:
rephrased it, let me know if it's better.

m-kuhn commented on tests/src/python/test_qgsfileutils.py:
```diff
@@ -92,6 +92,60 @@ def testFindClosestExistingPath(self):
+
+        #unexisting
+        files = QgsFileUtils.findFile(filename, nest, 1, 4)
+        print(files, '105')
```
I think you either want to remove the prints completely or replace them with something that gives more context than outdated line numbers ;)

roya0045 replied:
I just want to see if the test passes with the renaming that I did, this will help me find any mistakes on my end, once all tests are good I'll remove them.

m-kuhn replied:
ok, no worries

roya0045 replied:
Done, all tests are ok, forgot to sip, should be fine now.

roya0045 commented:
@m-kuhn Anything else or are you ok to merge this baby?

elpaso commented on python/core/auto_generated/qgsfileutils.sip.in:
```diff
@@ -84,6 +84,22 @@ and "/home/user/projects" exists but no "2018" subfolder exists, then the functi
+:param fileName: Name of the file to find
+:param basepath: current basepath of the file
+:param maxDepth: limit the number of time the search can move up
+:param driveMargin: limits how close the search can get to the drive level
```
A "drive"? I don't think it makes much sense for a *nix user.
So, first question is: is this a windows specific implementation?
If it is not (as I hope) can you rename this parameter and explain in the documentation of the argument what it exactly does?

roya0045 replied:
If you are willing to provide better terminology I can certainly change it. I was not too sure how to express this variable, but the gist is to maintain a buffer between the places that will be searched and the root. This is to prevent the alg from searching the full pc/server and contain it in a sub branch of the architecture.

I can call it root margin but I doubt this would help either.

roya0045 replied:
Please see the changes that I made, they should make things clearer, thank you for pushing me to change them!

elpaso commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -364,79 +372,65 @@ void QgsHandleBadLayers::editAuthCfg()
+    const QString fileType = mLayerList->item( i, 2 )->text();
+    if ( provider.toLower().toStdString() == "none" )
+    {
+      if ( fileType.toStdString() == "raster" )
```
Why converting to std::string? Would a straight comparison with `QStringLiteral( "raster" )` work?

roya0045 replied:
I just wanted to set it on the same base, either in qt or std c++. The conversion does not matter to me. I will change it.

roya0045 replied:
My merge squash broke the review but this was addressed.

elpaso reviewed (changes requested):
Looks good to me, thank you!
I just left a couple of comments.

roya0045 commented:
> Looks good to me, thank you!
> I just left a couple of comments.

Thank a lot for the review!

roya0045 commented:
@elpaso @m-kuhn I reworked the docs to make things more clear and added a functionality to pass the full path instead of the two parts. Let me know if you are ok with these changes.

roya0045 commented:
Test failure is in the svg cache, not related to this PR.

m-kuhn commented on src/core/qgsfileutils.h:
```diff
@@ -87,6 +87,19 @@ class CORE_EXPORT QgsFileUtils
+     * \returns List of strings of the first matching path in unix format.
+     * \since QGIS 3.12
+     */
+    static QStringList findFile( const QString &file, const QString &basepath = "", int maxClimbs = 4, int searchCeilling = 4, const QString &currentDir = "" );
```
```suggestion
    static QStringList findFile( const QString &file, const QString &basepath = QString(), int maxClimbs = 4, int searchCeilling = 4, const QString &currentDir = QString() );
```

elpaso replied:
also s/Ceilling/Ceiling

roya0045 replied:
I'll rebase in the hope that this solves the svg tests. These changes should be included.

roya0045 replied:
Changes are implemented and we have a green arrow.

roya0045 committed:
auto layer finder

roya0045 commented:
@m-kuhn @nyalldawson @elpaso Thank you for the reviews and the merge!
